window-icons/window-icons.patch

```diff
@@ -1,198 +1,197 @@
 diff --git a/include/data.h b/include/data.h
-index 58e4a00..b6c412f 100644
+index 3a059e7..d1f09ec 100644
 --- a/include/data.h
 +++ b/include/data.h
-@@ -421,6 +421,11 @@ struct Window {
+@@ -426,6 +426,9 @@ struct Window {
  
      /* aspect ratio from WM_NORMAL_HINTS (MPlayer uses this for example) */
      double aspect_ratio;
 +
-+#ifdef USE_ICONS
 +    /** Window icon, array of size 16x16 containing the ARGB pixels */
 +    uint32_t* icon;
-+#endif
  };
  
  /**
+diff --git a/include/libi3.h b/include/libi3.h
+index 4c72267..1d40cea 100644
+--- a/include/libi3.h
++++ b/include/libi3.h
+@@ -587,6 +587,11 @@ color_t draw_util_hex_to_color(const char *color);
+ void draw_util_text(i3String *text, surface_t *surface, color_t fg_color, color_t bg_color, int x, int y, int max_width);
+ 
+ /**
++ * Draw the given image using libi3.
++ */
++void draw_util_image(unsigned char *pixels, surface_t *surface, int x, int y, int width, int height);
++
++/**
+  * Draws a filled rectangle.
+  * This function is a convenience wrapper and takes care of flushing the
+  * surface as well as restoring the cairo state.
 diff --git a/include/window.h b/include/window.h
-index 395c988..f7a928d 100644
+index d0b97f1..0b2997d 100644
 --- a/include/window.h
 +++ b/include/window.h
-@@ -68,6 +68,15 @@ void window_update_type(i3Window *window, xcb_get_property_reply_t *reply);
+@@ -87,3 +87,9 @@ void window_update_hints(i3Window *win, xcb_get_property_reply_t *prop, bool *ur
+  *
   */
- void window_update_hints(i3Window *win, xcb_get_property_reply_t *prop, bool *urgency_hint);
- 
+ void window_update_motif_hints(i3Window *win, xcb_get_property_reply_t *prop, border_style_t *motif_border_style);
 +
-+#ifdef USE_ICONS
 +/**
 + * Updates the _NET_WM_ICON
 + *
 + */
 +void window_update_icon(i3Window *win, xcb_get_property_reply_t *prop);
+diff --git a/libi3/draw_util.c b/libi3/draw_util.c
+index dcd881c..3bc7ff9 100644
+--- a/libi3/draw_util.c
++++ b/libi3/draw_util.c
+@@ -153,6 +153,39 @@ void draw_util_text(i3String *text, surface_t *surface, color_t fg_color, color_
+ #endif
+ }
+ 
++
++/**
++ * Draw the given image using libi3.
++ * This function is a convenience wrapper and takes care of flushing the
++ * surface as well as restoring the cairo state.
++ *
++ */
++void draw_util_image(unsigned char *pixels, surface_t *surface, int x, int y, int width, int height) {
++    RETURN_UNLESS_SURFACE_INITIALIZED(surface);
++
++#if CAIRO_SUPPORT
++    cairo_save(surface->cr);
++
++    cairo_surface_t *image;
++
++    image = cairo_image_surface_create_for_data(
++            pixels,
++            CAIRO_FORMAT_ARGB32,
++            width,
++            height,
++            width * 4);
++
++    cairo_set_operator(surface->cr, CAIRO_OPERATOR_OVER);
++    cairo_set_source_surface(surface->cr, image, x, y);
++    cairo_rectangle(surface->cr, x, y, width, height);
++    cairo_fill(surface->cr);
++
++    cairo_surface_destroy(image);
++
++    cairo_restore(surface->cr);
 +#endif
```

Comment:
If I understand correctly, compiling without cairo support results in no icons being drawn. It might be worthwhile to add a simple README for the patch that documents this fact to avoid confusion (and corresponding issues being raised).